### ISSUE TITLE
The data on context map is lost when using resilience4j with Bulkhead & Retry

### ISSUE THREAD

**Issue by ioda01** [resilience4j/resilience4j#1476]
Resilience4j version: 1.7.1

Java version: 11

Hi Robert,

I am trying to integrate resilience4j on a service and I want to use the resilience4j circuit breaker with retry, bulkead threadpool and timelimiter with annotations. For the bulkhead I configured a contextPropagator that just copies the MDC context and clears the MDC in the clear implementation. This works perfectly in case there's no retry. In the situation a retry is triggered MDC is empty because after the first attempt the MDC is cleared by the MDCContextPropagator and when the ContextAwareScheduledThreadPoolExecutor tries to copy the context it's already empty.

Is there another way to use this?

This is the way I am using the features:

`
@CircuitBreaker(name = "Inst1", fallback="fallback")
@Retry(name = "Inst1")
@Bulkhead(name = "Inst1", type= Type.THREADPOOL)
@TimeLimiter(name = "Inst1", fallback = "fallback")
public CompletableFuture<MyResponse> request(....) {
  .....
}
`
My configurations are like this:

`
resilience4j:
  scheduled:
    executor:
      corePoolSize: 10
resilience4j.circuitbreaker:
  configs:
    default:
      slidingWindowSize: 10
      slidingWindowType: TIME_BASED
      permittedNumberOfCallsInHalfOpenState: 10
      waitDurationInOpenState: 5000
      failureRateThreshold: 50
      minimumNumberOfCalls: 20
      recordExceptions:
        - my.MyException
      ignoreExceptions:
        - org.springframework.web.server.ResponseStatusException
  instances:
    Inst1:
      baseConfig: default
resilience4j.retry:
  configs:
    default:
      maxRetryAttempts: 3
      waitDuration: 100
      retryExceptions:
        - my.MyException
  instances:
    Inst1:
      baseConfig: default
resilience4j.thread-pool-bulkhead:
  configs:
    default:
      maxThreadPoolSize: 50
      coreThreadPoolSize: 25
      queueCapacity: 25
      keepAliveDuration: 25
      contextPropagators:
        - my.MDCContextPropagator
  instances:
    Inst1:
      baseConfig: default
resilience4j.timelimiter:
  configs:
    default:
      cancelRunningFuture: true
      timeoutDuration: 10000
  instances:
    Inst1:
      baseConfig: default
`

**Comment by RobWin** [comment]
Hi,

try this:

```
resilience4j.scheduled.executor:
  corePoolSize: 10
  contextPropagators:
    - my.MDCContextPropagator
```

**Comment by ioda01** [comment]
If I move the contextPropagotors from bulkead level to exceturor level it's the context is not propagated at all...not even for first call.

**Comment by RobWin** [comment]
Ok, I see.
Your issue is related to this one: 

https://github.com/resilience4j/resilience4j/issues/1417

I think there is no solution right now. @mattca Did you have time to look at it?

**Comment by ioda01** [comment]
Yes, I think it's the same situation. Didn't knew about the issue from @mattca.

I think I found a workaround:

Instead of clearing the MDC in clear method of the ContextPropagator I just copy it again and leave the ContextAwareScheduledThreadPoolExecutor to clear it context.
I still have to see what are the consequences for this...

**Comment by EduGordSerasa** [comment]
Adding a `Resilience4JCircuitBreakerFactory` default customizer using `ContextAwareScheduledThreadPoolExecutor` worked for me as follows:

```java
    @Bean
    public Customizer<Resilience4JCircuitBreakerFactory> defaultCustomizer() {
        return factory -> {
            ContextAwareScheduledThreadPoolExecutor executor = ContextAwareScheduledThreadPoolExecutor
                    .newScheduledThreadPool()
                    .build();
            factory.configureExecutorService(executor);
        };
    }
```

for reactive users 

```java
    @Bean
    public Customizer<ReactiveResilience4JCircuitBreakerFactory> defaultCustomizer() {
        return factory -> {
            ContextAwareScheduledThreadPoolExecutor executor = ContextAwareScheduledThreadPoolExecutor.
                    newScheduledThreadPool()
                    .build();
            factory.configureExecutorService(executor);
        };
    }
```

reference: https://docs.spring.io/spring-cloud-circuitbreaker/reference/spring-cloud-circuitbreaker-resilience4j/default-configuration.html
